### PR TITLE
adapted linopt for tranfer line

### PR DESCRIPTION
**Pull request by swhite2401**
Hello @lfarv 
I have implemented a potential adaptation of linopt for transfer lines.

-additional optional argument twiin to consider a transfer line

Also included in this branch:
-do not return unnecessary outputs for uncoupled lattice (speed-up)
-remove get_twiss()

Could you please take a look and let me know what you think?

(I haven't fixed the help and PEP8 yet)

Thanks!

### CODE REVIEW

File: pyat/at/physics/linear.py [hunk 1]
```diff
@@ -305,17 +217,25 @@ def analyze(r44):
...
+        a0_b, b0_b, tune_b = _closure(B)
+        tune = numpy.array([tune_a, tune_b])
+
+    if get_chrom:      
+        if twiin is not None:
+            orbit[0:5] += numpy.append(disp0*0.5*ddp, 0.5 * ddp)
         d0_up, tune_up, _, l_up = linopt(ring, dp + 0.5 * ddp, uintrefs,
                                          keep_lattice=True,
-                                         coupled=coupled)
+                                         coupled=coupled, twiin=twiin)
```
Comment (lfarv) [inline pyat/at/physics/linear.py:230]: You are changing the initial orbit `orbit`, but it is not used. Probably it should be introduced in `twiin`?

**Comment by swhite2401** [comment]
orbit is the pointer to twiin['closed_orbit'] so I think it is fine.
Am I missing something?

**Comment by lfarv** [comment]
OK, you are right, I missed it…
But to avoid confusion between `orbit` in `twiin` and the `orbit` argument, I would suggest taking `orbit` out of twiin and using the existing argument for both cases:
- for a ring: if orbit is none, orbit=findorbit4
- for a transfer line, if orbit is None, orbit is zero

Then for `get_chrom`, no need to increment orbit and to multiply the tests, it could be:
```
        if twiin is not None:
            dorbit = numpy.append(disp0*0.5*ddp, 0.5 * ddp)
            kwup = dict(orbit=orbit+dorbit, twiin=twiin)
            kwdown = dict(orbit=orbit-dorbit, twiin=twiin)
        else:
            kwup = {}
            kwdown = {}

        d0_up, tune_up, _, l_up = linopt(ring, dp=dp + 0.5, refpts=uintrefs,
                                         keep_lattice=True,
                                         coupled=coupled, **kwup)
        d0_down, tune_down, _, l_down = linopt(ring, dp - 0.5, refpts=uintrefs,
                                               keep_lattice=True,
                                               coupled=coupled, **kwdown)

        chrom = (tune_up - tune_down) / ddp
        dispersion = (l_up['closed_orbit'] -
                      l_down['closed_orbit'])[:, :4] / ddp
        disp0 = (d0_up['closed_orbit'] - d0_down['closed_orbit'])[:4] / ddp
```
I also have a doubt: `orbit` and `dorbit` should have dimension (6,). The append statement adds a fifth value to a (4,) dispersion. Is there a zero missing?

Finally, the same variable `disp0` is used 1st as the initial dispersion in `twiin` and then for the computed value at the beginning of the structure. One could avoid the first definition by defining the orbit displacement `dorbit` directly in the argument checking section, instead of `disp0`:

```
        try:      
            dorbit = numpy.append(0.5*ddp*twiin['dispersion'], 0.5 * ddp)
        except KeyError: 
            print('Dispersion not found in twiin, setting to zero')
            dorbit = numpy.append(numpy.zeros((4,)), 0.5 * ddp)
```

**Comment by willrogers** [comment]
Just a comment that you are free to ignore: there's a tendency to use very compact variable names and I find it makes it a bit more difficult to understand, partly because I am not as familiar with the physics.

e.g. is `twissin` or `twiss_in` not more understandable than `twiin`?

**Comment by swhite2401** [comment]
Thanks Laurent, I have implemented your suggestions.
Concerning orbit the answer is yes, this is why only `orbit[1:5]` was incremented in the previous version.

We could possibly remove additional tests by requesting the type `lindata` for twiin but thhis removes some flexibility

**Comment by swhite2401** [comment]
I agree `twiss_in` is better I'll change this now

**Comment by lfarv** [comment]
To get the tests passed, you need to remove the test on `get_twiss` in `test/test_physics.py`. But there is also a problem in test_linopt_uncoupled, where "gamma" for an uncoupled line should be 1 all along the line and not nan. But it looks really useless for an uncoupled line, so probably removing the test is OK.

**Comment by swhite2401** [comment]
Yes I was waiting for @willrogers  to give his green light to remove `get_twiss()` to start modifying `test_physics.py`

**Comment by willrogers** [comment]
I should do this testing but I can't guarantee I can do it promptly enough not to hold you up. So I suggest you delete `get_twiss()` and I can always put it back manually.

**Comment by swhite2401** [comment]
Ok then let me put back `get_twiss()` so that the checks pass and you have both functions to do your test.
Once you are satisfied we can decide to remote it or not.

**Comment by willrogers** [comment]
See comment on #171.

**Comment by swhite2401** [comment]
Ok I will remove get_twiss() as agreed and then this part is probably ready for merging

**Review by lfarv — APPROVED** [review]

**Comment by willrogers** [comment]
OK for me.

**Comment by willrogers** [comment]
I can resolve this conflict if you like.

**Comment by swhite2401** [comment]
I was still working on that one, I restore it and do the merge when ready

**Pull request merged**